### ISSUE TITLE
device parameter in yolo predict does not seem work when using exported onnx model

### ISSUE THREAD

**Issue by rabum** [ultralytics/ultralytics#22538]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I set device=1 but the model is still runned on device 0

**Comment by Y-T-G** [comment]
What's the full code you used?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @rabum, thanks for reporting this and for using Ultralytics 🚀 This is an automated response to help you get support faster — an Ultralytics engineer will follow up shortly.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many common questions may already be answered.

Since this appears related to inference with an exported ONNX model and the `device` parameter, if this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) so we can investigate. Helpful details include:
- Exact command(s) or code you’re running (show how `device=1` is set)
- `ultralytics` version and install method
- How the ONNX model was exported (command and model used)
- Environment details: OS, Python version, GPU(s), CUDA/CUDNN, PyTorch version
- Console logs and any warnings/errors
- Whether this occurs in both Python API and CLI, and with non-ONNX models

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/) to verify the issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Google Cloud Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- Amazon Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- Docker Image. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by rabum** [comment]
> What's the full code you used?

the original code is a bit too long to be posted here, the related part is :
```
class Model:
    def __init__(self, **kwargs):
        ......
    def activate_model(self):
        if self.type in ['yolo']:
            self.model = YOLO(self.path)
    def predict(self, batch_frame, **kwargs):
        if self.type in ['yolo']:
            if self.device == -1:
                results = self.model.predict(batch_frame, verbose=False)
            else:
                results = self.model.predict(batch_frame, verbose=False, device=self.device)
```
Basically I use another program to call this class for model inference. The problem is even when self.device=1, there is still only device 0 in use, only when I set os.environ["CUDA_VISIBLE_DEVICES"] = "1" at the top of the file device 1 can be correctly used, but I don't want to use this way since device is a parameter to be entered.

**Comment by glenn-jocher** [comment]
Thanks for the snippet—this is an ONNX-specific gap: the GPU index is fixed when the ONNX Runtime session is created, and our current `AutoBackend` doesn’t pass a `device_id`, so it defaults to GPU 0; the `device` arg in `predict()` won’t rebind an already-loaded ONNX model. See how ONNX is initialized in the backend in the [AutoBackend reference](https://docs.ultralytics.com/reference/nn/autobackend/), and the `device` arg scope in the [prediction guide](https://docs.ultralytics.com/modes/predict/).

Workarounds for now: set `CUDA_VISIBLE_DEVICES` before constructing `YOLO` (as you found), or create the ORT session yourself with a device id:
```python
import onnxruntime as ort
sess = ort.InferenceSession(
    "model.onnx",
    providers=[('CUDAExecutionProvider', {'device_id': 1}), 'CPUExecutionProvider']
)
```

Action we can take: add provider options in `ultralytics/nn/autobackend.py` so `device=1` maps to ORT GPU 1, e.g.:
```python
# inside ONNX branch
providers = [('CUDAExecutionProvider', {'device_id': device.index})] if cuda else ['CPUExecutionProvider']
session = onnxruntime.InferenceSession(w, providers=providers)
```
If you can share the output of `yolo checks` and your `onnxruntime`/`onnxruntime-gpu` versions, that will help us validate across environments; the checks tool is documented here: [check_yolo reference](https://docs.ultralytics.com/reference/utils/checks/).

**Comment by rabum** [comment]
> Thanks for the snippet—this is an ONNX-specific gap: the GPU index is fixed when the ONNX Runtime session is created, and our current `AutoBackend` doesn’t pass a `device_id`, so it defaults to GPU 0; the `device` arg in `predict()` won’t rebind an already-loaded ONNX model. See how ONNX is initialized in the backend in the [AutoBackend reference](https://docs.ultralytics.com/reference/nn/autobackend/), and the `device` arg scope in the [prediction guide](https://docs.ultralytics.com/modes/predict/).
> 
> Workarounds for now: set `CUDA_VISIBLE_DEVICES` before constructing `YOLO` (as you found), or create the ORT session yourself with a device id:
> 
> import onnxruntime as ort
> sess = ort.InferenceSession(
>     "model.onnx",
>     providers=[('CUDAExecutionProvider', {'device_id': 1}), 'CPUExecutionProvider']
> )
> Action we can take: add provider options in `ultralytics/nn/autobackend.py` so `device=1` maps to ORT GPU 1, e.g.:
> 
> # inside ONNX branch
> providers = [('CUDAExecutionProvider', {'device_id': device.index})] if cuda else ['CPUExecutionProvider']
> session = onnxruntime.InferenceSession(w, providers=providers)
> If you can share the output of `yolo checks` and your `onnxruntime`/`onnxruntime-gpu` versions, that will help us validate across environments; the checks tool is documented here: [check_yolo reference](https://docs.ultralytics.com/reference/utils/checks/).

parse_requirements: [namespace(name='numpy', specifier='<=2.1.1,>=1.23.0'), namespace(name='matplotlib', specifier='>=3.3.0'), namespace(name='opencv-python', specifier='>=4.6.0'), namespace(name='pillow', specifier='>=7.1.2'), namespace(name='pyyaml', specifier='>=5.3.1'), namespace(name='requests', specifier='>=2.23.0'), namespace(name='scipy', specifier='>=1.4.1'), namespace(name='torch', specifier='>=1.8.0'), namespace(name='torch', specifier='!=2.4.0,>=1.8.0; sys_platform == "win32"'), namespace(name='torchvision', specifier='>=0.9.0'), namespace(name='tqdm', specifier='>=4.64.0'), namespace(name='psutil', specifier=''), namespace(name='py-cpuinfo', specifier=''), namespace(name='pandas', specifier='>=1.1.4'), namespace(name='seaborn', specifier='>=0.11.0'), namespace(name='ultralytics-thop', specifier='>=2.0.0')]

pip show onnxruntime-gpu
Name: onnxruntime-gpu
Version: 1.23.2
Summary: ONNX Runtime is a runtime accelerator for Machine Learning models
Home-page: https://onnxruntime.ai
Author: Microsoft Corporation
Author-email: onnxruntime@microsoft.com
License: MIT License
Location: /home/yunzheng/anaconda3/lib/python3.12/site-packages
Requires: coloredlogs, flatbuffers, numpy, packaging, protobuf, sympy
Required-by:

**Comment by glenn-jocher** [comment]
Thanks for the details — `onnxruntime-gpu 1.23.2` supports per-GPU selection, so a small fix in `AutoBackend` will map `device=1` to ORT GPU 1. If you’d like to try locally, update the ONNX branch to pass provider options:

```python
# ultralytics/nn/autobackend.py (ONNX branch)
import onnxruntime
providers = ["CPUExecutionProvider"]
if cuda and "CUDAExecutionProvider" in onnxruntime.get_available_providers():
    gpu_id = device.index if isinstance(device, torch.device) and device.index is not None else 0
    providers = [("CUDAExecutionProvider", {"device_id": gpu_id}), "CPUExecutionProvider"]
session = onnxruntime.InferenceSession(w, providers=providers)
```

Please also paste the full text output of `yolo checks` and the result of:
```python
import onnxruntime as ort
print(ort.get_available_providers())
```
so we can validate on your setup. For background on ONNX GPU execution with YOLO, see the ONNX integration guide: [ONNX export for YOLO11 models](https://docs.ultralytics.com/integrations/onnx/). If you’re up for it, a minimal PR with this change and a small test would be welcome.

**Comment by UltralyticsAssistant** [comment]
Great news — a potential fix has been merged in the pull request Specify device ID for ONNX inference on CUDA: https://github.com/ultralytics/ultralytics/pull/22546.

What changed and why this may resolve the issue:
- ONNX Runtime is now initialized with explicit CUDA provider options, passing the selected GPU’s device_id to `CUDAExecutionProvider`. This ensures inference respects the chosen device (e.g., `cuda:1`) instead of defaulting to GPU 0.
- If CUDA isn’t available, behavior remains the same: a warning is shown and inference falls back to CPU.
- No API changes are required; existing ONNX inference code benefits automatically when a device is specified.

Huge thanks to @Y-T-G for the contribution and to everyone who helped validate multi-GPU behavior.

How to test the fix:
- Upgrade to the latest Ultralytics package when the next release is published on PyPI:
  - `pip install -U ultralytics`
- Or install directly from the main branch to test immediately:
  - `pip install -U git+https://github.com/ultralytics/ultralytics.git@main`
- Ensure ONNX Runtime is up to date (GPU builds for CUDA):
  - `pip install -U onnxruntime-gpu`
- Test specifying a non-default GPU:
  - Python:
    ```python
    from ultralytics import YOLO

    model = YOLO("path/to/model.onnx")
    results = model.predict(source="path/to/video_or_image.jpg", device="cuda:1")
    ```
  - CLI:
    ```
    yolo predict model=path/to/model.onnx source=path/to/video_or_image.jpg device=cuda:1
    ```

Feedback request:
- Please try the above and share whether the merged changes resolve the issue in your multi-GPU environment.
- If problems persist, details such as OS, CUDA/driver versions, `onnxruntime(-gpu)` version, and a brief repro snippet would be very helpful.

Thanks for reporting this and helping improve ONNX Runtime inference reliability. 🙏
